Add Symfony 5 compatibility

tomvlk opened the pull request:
Changes in the configuration loader will add support for Symfony 5. Of course this might be breaking for other versions so it must be a new major version.

tomvlk committed:
Symfony 5 compatibility

tomvlk committed:
Update twig extension to be compatible with Symfony 5 and new Twig versions

lifo101 commented on DependencyInjection/Configuration.php:
```diff
@@ -18,8 +18,8 @@ class Configuration implements ConfigurationInterface
     {
-        $treeBuilder = new TreeBuilder();
-        $rootNode = $treeBuilder->root('lifo_typeahead');
+        $treeBuilder = new TreeBuilder('lifo_typeahead');
+        $rootNode = $treeBuilder->getRootNode();
```
If you do that you'll lose backwards compatibility. I haven't gotten around to doing this to typeahead, but I do it in other projects:

```php
        if (Kernel::VERSION_ID < 40000) {
            $treeBuilder = new TreeBuilder();
            $rootNode = $treeBuilder->root('lifo_typeahead');
        } else {
            $treeBuilder = new TreeBuilder('lifo_typeahead');
            $rootNode = $treeBuilder->getRootNode();
        }
```

tomvlk committed:
Adjusting the configuration to support Symfony 5 and to keep backward compatibility

tomvlk commented:
Correct, solved it. It needs extensive testing, something that I haven't done at the moment.

tomvlk committed:
Correctly fix the last commit with using an use statement...

tomvlk commented:
Btw, don't merge this yet as I will add more changes most likely once I get a project migrated.

tomvlk commented:
@lifo101 I still have this issue right now and I have no idea how to solve it other than manual fixing it in the twig configuration of my installation

```
Unable to find template "LifoTypeaheadBundle:Form:typeahead.html.twig" (looked into: /code/templates, /code/vendor/symfony/twig-bridge/Resources/views/Form).
```

Fixing it manually like this solves it. (it introduces other bugs which I will fix and add to the pull request btw).
Manual Fix in twig.yaml:
```
paths:
    '%kernel.project_dir%/vendor/lifo/typeahead-bundle/Lifo/Resources/views': LifoTypeaheadBundle
```
Manual Fix in the typeahead bundle extension (LifoTypeaheadExtension.php at the configureTwigBundle function:
```
            $resources = array('LifoTypeaheadBundle:Form:typeahead.html.twig');
            if (Kernel::VERSION_ID >= 40000) {
                $resources = array('@LifoTypeaheadBundle/Form/typeahead.html.twig');
                $container->prependExtensionConfig('twig', array('form_themes' => $resources));
            } elseif (Kernel::VERSION_ID >= 20600) {
                $container->prependExtensionConfig('twig', array('form_themes' => $resources));
            } else {
                $container->prependExtensionConfig('twig', array('form' => array('resources' => $resources)));
            }
```

tomvlk committed:
Make sure it loads the correct namespaced template file in Symfony 4 and up.

tomvlk committed:
Correct the resource file to load in symfony 4 and later.

tomvlk committed:
Update form template to support new twig syntax (but with backward compatibility)

Unfortunately losing the cleanness but I had to remove the nested for .. if for compatibility issues, and the 'spaceless' filters as it's now applied differently which is not backward compatible.

tomvlk commented:
@lifo101 I got it working in Symfony 5 right now. It still needs testing in different 4.x versions and < 4.0 versions. It would be nice if you could help with that maybe, and give maybe additional feedback. I had to remove several things from the template like spaceless filter (as it's now implemented differently in twig, but unable to apply this due to backward compatibility reasons). And I had to undo the nested for .. if clause as it's now working with filters in the new Twig versions.

benstinton commented:
@lifo101 thank you very much for typeahead-bundle, have used it lots over the years. Looking forward to https://github.com/lifo101/autocomplete-bundle , will definitely check it out as soon as you have tagged a release.
In the mean time it would be great to eek a little bit more out of typeahead-bundle with this PR if it works.
@tomvlk great work, I'd be more than happy to test this on a SF4 build with diverse uses of TypeaheadType. 

Any feedback as soon as would be appreciated, as working on an SF4 upgrade right now and having to decide wether to rip out typeahead or not.

lifo101 commented:
I won't be able to do much with this due to the holidays and vacation until after the new year.

benstinton commented:
Just for feedback, I've now got this branch deployed in production environments as it working well in Symfony 4.4

tomvlk commented:
Would be nice if this can get into the official release soon. 
Thanks all for the work on the repository.
